### ISSUE TITLE
pythonw.exe

### ISSUE THREAD

**Issue by jpmassena** [wakatime/jetbrains-wakatime#65]
Hi, I have Android Studio 2.0 Beta 5 and everytime I try to debug or build the project, I get dozens (maybe more than hundred) of processes pythonw.exe from wakatime, each with ~20Mb of ram to the point where my cpu reaches 100% during 5 minutes.

The relevant logs I get are those:

```
{"now": "2016/02/24 16:16:18 -0000", "version": "4.1.10", "plugin": "AndroidStudio/2.0 Beta 5 AndroidStudio-wakatime/6.0.5", "time": 1456330487.5204053, "level": "ERROR", "message": "Traceback (most recent call last):\n  File \"...\\.AndroidStudioPreview2.0\\config\\plugins\\WakaTime-resources\\wakatime-master\\wakatime\\session_cache.py\", line 52, in save\n    c.execute('DELETE FROM session')\nsqlite3.OperationalError: database is locked\n"}

{"now": "2016/02/24 16:16:22 -0000", "version": "4.1.10", "plugin": "AndroidStudio/2.0 Beta 5 AndroidStudio-wakatime/6.0.5", "time": 1456330487.9214282, "level": "ERROR", "message": "Traceback (most recent call last):\n  File \"...\\.AndroidStudioPreview2.0\\config\\plugins\\WakaTime-resources\\wakatime-master\\wakatime\\session_cache.py\", line 80, in get\n    c.execute('BEGIN IMMEDIATE')\nsqlite3.OperationalError: database is locked\n"}

```

I get lots of those lines.

**Comment by alanhamlett** [comment]
That log line is a symptom of the problem, it must be trying to send too many heartbeats at once. It's probably related to this line https://github.com/wakatime/jetbrains-wakatime/blob/f7591c39da7b1725a737d9a9cbb3c92ecf3085a1/src/com/wakatime/intellij/plugin/CustomDocumentListener.java#L29

Can you turn on debug mode in Android Studio from Tools -> WakaTime Settings menu then reproduce this problem and send me a copy of your `~/.wakatime.log` file via email alan@wakatime.com?

**Comment by jpmassena** [comment]
Hi, just sent you an email (two actually, forgot to include the logfile in the first one) with the info you requested.

**Comment by alanhamlett** [comment]
Thanks for the logs!

You're opening an Activity and Android Studio translates that into editing those multiple xml files, but you aren't actually opening those xml files right? So ignoring files from `.gitignore` would work if using Git, but if someone names their build folders different than `build` and isn't using revision control.... maybe we can distinguish between `documentChanged` events originating from editing Activities (made up of multiple xml files) verses editing single code files?

**Comment by jpmassena** [comment]
Those logs (and pythonw.exe processes) only happen when I click the "debug" button, to test the app. When I do that, Android Studio builds the project and runs the executable on the phone/emulator.

All those files are `documentChanged` because Gradle cleans the `build `folder (deletes all generated files) and rebuilds them from the new sources.

The `build `folders are generated by the IDE, if someone changes it, they have to change the IDE settings (if it is actually possible, I don't know).

The `gitignore `files are also generated when starting a new project, even without source control, but those are easily deleted by the users.

My guess is that you could focus on ignoring the `build` folders.

**Comment by delissonjunio** [comment]
Just experienced this, wakatime would spawn pythonw.exe processes until my computer halted

**Comment by alanhamlett** [comment]
The new `--extra-heartbeats` argument tells wakatime-cli to read from STDIN a JSON array of heartbeat data. Now we can queue up heartbeats and send them once every few seconds using only one python process instead of forking many python processes.

Anyone want to take a stab at implementing this?

**Comment by alanhamlett** [comment]
Fixed with c394dbbc2eb14150a126c594c3f6bc9099864279 and released in v7.0.0.
